src/m365/commands/request.ts

```diff
@@ -0,0 +1,163 @@
+        resource: args.options.resource,
+        accept: args.options.accept || 'application/json',
+        contentType: args.options.contentType,
+        body: typeof args.options.body !== 'undefined',
+        headers: typeof args.options.headers !== 'undefined'
+      });
+    });
+  }
+
+  #initOptions(): void {
+    this.options.unshift(
+      {
+        option: '-u, --url <url>'
+      },
+      {
+        option: '-m, --method [method]',
+        autocomplete: ['get', 'post', 'put', 'patch', 'delete', 'head', 'options']
+      },
+      {
+        option: '-r, --resource [resource]'
+      },
+      {
+        option: '-b, --body [body]'
+      },
+      {
+        option: '-h, --headers [headers]'
+      },
+      {
+        option: '-a, --accept [accept]'
+      },
+      {
+        option: '-c, --contentType [contentType]'
+      }
+    );
+  }
+
+  #initValidators(): void {
+    this.validators.push(
+      async (args: CommandArgs) => {
+        if (args.options.body && (!args.options.method || args.options.method === "get")) {
+          return "Specify a different method when using body";
+        }
+
+        if (args.options.body && !args.options.contentType) {
+          return "Specify the contentType when using body";
+        }
+
+        return true;
+      }
+    );
+  }
+
+  public commandAction(logger: Logger, args: CommandArgs, cb: (err?: any) => void): void {
+    this.getRequestDetails(logger, args)
+      .then(details => this.executeRequest(logger, details))
+      .then(response => {
+        if (response && response !== '') {
+          if (!args.options.accept || args.options.accept.startsWith("application/json")) {
+            logger.log(JSON.parse(response));
+          }
+          else {
+            logger.log(response);
+          }
+        }
+        cb();
+      }, (rawRes: any): void => this.handleRejectedODataJsonPromise(rawRes, logger, cb));
+  }
+
+  private getRequestDetails(logger: Logger, args: CommandArgs): Promise<any> {
+    if (this.verbose) {
+      logger.logToStderr(`Preparing request...`);
+    }
+
+    const headers: AxiosRequestHeaders = args.options.headers ? JSON.parse(args.options.headers) : {};
+    const method = (args.options.method || 'get').toUpperCase();
+
+    if (args.options.accept || !headers['accept']) {
+      headers['accept'] = args.options.accept || "application/json";
+    }
+
+    if (args.options.contentType) {
+      headers['content-type'] = args.options.contentType;
+    }
+
+    const requestDetails: AxiosRequestConfig<string> = {
+      url: args.options.url,
+      headers,
+      method,
+      data: args.options.body
+    };
+
+    if (args.options.resource) {
+      if (this.verbose) {
+        logger.logToStderr(`Retrieving access token for resource...`);
+      }
+
+      const resource: string = Auth.getResourceFromUrl(args.options.resource);
+      return auth
+        .ensureAccessToken(resource, logger as Logger, this.debug)
```

Comment:
Since we already have logic in place to retrieve access token inside the request, I wonder if it wouldn't be better to pass the resource on the request config and retrieve the token there instead of repeating the token retrieval code here?

We could do something like `x-resource: resource` on the headers, in the request object check for it and use it as a resource or use the default resource detection logic if no resource was specified.

Removing this logic from here, will also allow us to let people use CLI to execute anonymous requests should they want to for consistency.

Reply:
done!